### ISSUE TITLE
flux_rpc_get_unpack on empty string returns ENOMEM

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#4413]
@grondo noted in #4409:

> I can't even figure out how to create an empty eventlog. Any hints? If I use flux_kvs_txn_put (txn, 0, key, "") then I get
```
$ flux mini run hostname
2022-07-17T02:25:18.141047Z job-info.err[0]: get_main_eventlog_continuation: flux_rpc_get_unpack: Cannot allocate memory
0.050s: job.exception type=exec severity=0 flux_job_event_watch_get: Cannot allocate memory
flux-job: task(s) exited with exit code 1
0.049s: flux-shell[0]: FATAL: input: flux_job_event_watch_get: Cannot allocate memory
````
> ENOMEM from flux_rpc_get_unpack(3) seems like another bug.

The `flux_kvs_txn_put ("")` _should_ create a kvs value containing an empty string (e.g. just a NULL character).

The `flux_rpc_get_unpack()` in `flux_job_event_watch_get()` attempts to unpack the format string `{s:s}` and fails appropriately, but the ENOMEM error is wrong.

**Comment by garlick** [comment]
Note, this works:
```
$ FLUX_HANDLE_TRACE=1 flux kvs put a=""
--------------------------------------
> 0.00000
> kvs.commit
> flags=topic,payload,route userid=unknown rolemask=none nodeid=any matchtag=1
> {"namespace":"primary","flags":0,"ops":[{"key":"a","flags":0,"dirent":{"ver":1,"type":"val","data":""}}]}
--------------------------------------
< 0.00042
< kvs.commit
< flags=topic,payload,route userid=5588 rolemask=owner errnum=0 matchtag=1
< {"rootref":"sha1-7880715848f056a10f0168c7ebd1febf46d6a846","rootseq":7}

$ FLUX_HANDLE_TRACE=1 flux kvs get a
--------------------------------------
> 0.00000
> kvs.lookup
> flags=topic,payload,route userid=unknown rolemask=none nodeid=any matchtag=1
> {"key":"a","namespace":"primary","flags":0}
--------------------------------------
< 0.00029
< kvs.lookup
< flags=topic,payload,route userid=5588 rolemask=owner errnum=0 matchtag=1
< {"val":{"data":"","type":"val","ver":1}}
```

**Comment by grondo** [comment]
This simpler reproducer shows that the problem might be in the job-info module:

```
$ FLUX_HANDLE_TRACE=1 flux kvs put job.0000.001b.7600.0000.a=
--------------------------------------
> 0.00000
> kvs.commit
> flags=topic,payload,route userid=unknown rolemask=none nodeid=any matchtag=1
> {"namespace":"primary","flags":0,"ops":[{"key":"job.0000.001b.7600.0000.a","flags":0,"dirent":{"ver":1,"type":"val","data":""}}]}
--------------------------------------
< 0.00144
< kvs.commit
< flags=topic,payload,route userid=1000 rolemask=owner errnum=0 matchtag=1
< {"rootref":"sha1-53f19d98e65adcdc4ba1c8fc2bb1dafc5b34f7df","rootseq":20}
$ FLUX_HANDLE_TRACE=1 flux job info 0000.001b.7600.0000 a
--------------------------------------
> 0.00000
> job-info.lookup
> flags=topic,payload,route userid=unknown rolemask=none nodeid=any matchtag=1
> {"id":117943828480,"keys":["a"],"flags":0}
--------------------------------------
< 0.00186
< job-info.lookup
< flags=topic,route userid=1000 rolemask=owner errnum=12 matchtag=1
flux-job: flux_rpc_get_unpack: Cannot allocate memory
```

**Comment by grondo** [comment]
Here's a trace from the `job-info` module during the above failure:
```
$ flux job info job.0000.0016.2f00.0000 a
--------------------------------------
> 43.40499
> job-info.lookup
> flags=topic,payload,route userid=1000 rolemask=owner nodeid=any matchtag=1
> |bda5698a!c2123864!099ffbfd|
> {"id":95277809664,"keys":["a"],"flags":0}
--------------------------------------
> 43.40510
> kvs.lookup
> flags=topic,payload,route userid=unknown rolemask=none nodeid=any matchtag=1
> {"key":"job.0000.0016.2f00.0000.a","namespace":"primary","flags":0}
--------------------------------------
< 43.40600
< kvs.lookup
< flags=topic,payload,route userid=1000 rolemask=owner errnum=0 matchtag=1
< {"val":{"data":"","type":"val","ver":1}}
--------------------------------------
< 43.40608
< job-info.lookup
< flags=topic,route userid=unknown rolemask=none errnum=12 matchtag=1
< |bda5698a!c2123864!099ffbfd|
flux-job: flux_rpc_get_unpack: Cannot allocate memory
--------------------------------------
> 43.40669
> job-info.disconnect
> flags=topic,noresponse,route userid=1000 rolemask=owner nodeid=any matchtag=1
> |bda5698a!c2123864!099ffbfd|
```

**Comment by grondo** [comment]
The issue is in `info_lookup_continuation()`, an empty value is returned as NULL and `json_string()` fails, which is translated to `ENOMEM`:

```
diff --git a/src/modules/job-info/lookup.c b/src/modules/job-info/lookup.c
index c8ce65713..95094b245 100644
--- a/src/modules/job-info/lookup.c
+++ b/src/modules/job-info/lookup.c
@@ -204,7 +206,7 @@ static void info_lookup_continuation (flux_future_t *fall, void *arg)
             goto error;
         }
 
-        if (!(str = json_string (s)))
+        if (!(str = json_string (s?s:"")))
             goto enomem;
 
         if (json_object_set_new (o, keystr, str) < 0) {
```

But after this quick fix, the broker later segfaults due to the empty eventlog:

```
$ flux mini run hostname
==4529== Thread 12:
==4529== Invalid read of size 1
==4529==    at 0x61A0E60: __strchr_avx2 (strchr-avx2.S:45)
==4529==    by 0xA16E83C: eventlog_parse_next (watch.c:197)
==4529==    by 0xA16E83C: watch_continuation (watch.c:268)
==4529==    by 0x4E8A302: ev_invoke_pending (ev.c:3770)
==4529==    by 0x4E8E058: ev_run (ev.c:4190)
==4529==    by 0x4E5B1CA: flux_reactor_run (reactor.c:128)
==4529==    by 0xA16D79E: mod_main (job-info.c:136)
==4529==    by 0x117778: module_thread (module.c:183)
==4529==    by 0x50CE6DA: start_thread (pthread_create.c:463)
==4529==    by 0x613461E: clone (clone.S:95)
==4529==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==4529== 
==4529== 
==4529== Process terminating with default action of signal 11 (SIGSEGV)
==4529==  Access not within mapped region at address 0x0
==4529==    at 0x61A0E60: __strchr_avx2 (strchr-avx2.S:45)
==4529==    by 0xA16E83C: eventlog_parse_next (watch.c:197)
==4529==    by 0xA16E83C: watch_continuation (watch.c:268)
==4529==    by 0x4E8A302: ev_invoke_pending (ev.c:3770)
==4529==    by 0x4E8E058: ev_run (ev.c:4190)
==4529==    by 0x4E5B1CA: flux_reactor_run (reactor.c:128)
==4529==    by 0xA16D79E: mod_main (job-info.c:136)
==4529==    by 0x117778: module_thread (module.c:183)
==4529==    by 0x50CE6DA: start_thread (pthread_create.c:463)
==4529==    by 0x613461E: clone (clone.S:95)
==4529==  If you believe this happened as a result of a stack
==4529==  overflow in your program's main thread (unlikely but
==4529==  possible), you can try to increase the size of the
==4529==  main thread stack using the --main-stacksize= flag.
==4529==  The main thread stack size used in this run was 8388608.
==4529== 
==4529== HEAP SUMMARY:
==4529==     in use at exit: 7,462,142 bytes in 15,404 blocks
==4529==   total heap usage: 72,569 allocs, 57,165 frees, 5,521,641,700 bytes allocated
==4529== 
==4529== LEAK SUMMARY:
==4529==    definitely lost: 88 bytes in 1 blocks
==4529==    indirectly lost: 0 bytes in 0 blocks
==4529==      possibly lost: 4,249,520 bytes in 31 blocks
==4529==    still reachable: 3,212,534 bytes in 15,372 blocks
==4529==                       of which reachable via heuristic:
==4529==                         length64           : 209,400 bytes in 228 blocks
==4529==         suppressed: 0 bytes in 0 blocks
==4529== Rerun with --leak-check=full to see details of leaked memory
==4529== 
==4529== For counts of detected and suppressed errors, rerun with: -v
==4529== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
Segmentation fault (core dumped)
```

**Comment by chu11** [comment]
Ahhh @grondo beat me to it :-)  I suppose this is the "fallout" I alluded to in #4409 about if we would want to allow empty eventlogs.

Clearly some corner case checks should be added here to ensure the user gets a clean error message.  But independent of that (and related to discussion in #4409), should an empty eventlog be allowed?

My initial gut feeling is we don't.  This is one set of corner cases, but I imagine there would be many other corner cases throughout flux if we open up the can of worms where exec.eventlog can be empty, input/output can be empty, and perhaps other things too.

**Comment by garlick** [comment]
My vote is we use the fix proposed in #4412 and let the RFC 18 rule forbidding empty eventlogs stand.

But IMHO we should still fix the job-info error and make libeventlog handle invalid inputs.  I bet a user can crash the broker by creating an empty eventlog in the guest namespace and asking job-info to read it.
